_posts/2020-10-05-torch-network-with-autograd/torch-network-with-autograd.Rmd

```diff
@@ -0,0 +1,347 @@
+
+z <- y$pow(2) * 3
+
+out <- z$mean()
+```
+
+To save memory, by default intermediate gradients are not stored.
+Calling `retain_grad()` on a tensor allows to deviate from this default;
+let's do this here, for the sake of demonstration:
+
+```{r}
+y$retain_grad()
+
+z$retain_grad()
+```
+
+Now we can go backwards through the graph and inspect `torch`'s action
+plan for backprop, starting from `out$grad_fn`, like so:
+
+```{r}
+# how to compute the gradient for mean, the last operation executed
+out$grad_fn
+```
+
+    MeanBackward0
+
+```{r}
+# how to compute the gradient for the multiplication by 3 in z = y.pow(2) * 3
+out$grad_fn$next_functions
+```
+
+    [[1]]
+    MulBackward1
+
+```{r}
+# how to compute the gradient for pow in z = y.pow(2) * 3
+out$grad_fn$next_functions[[1]]$next_functions
+
+```
+
+    [[1]]
+    PowBackward0
+
+```{r}
+# how to compute the gradient for the multiplication in y = x * (x + 2)
+out$grad_fn$next_functions[[1]]$next_functions[[1]]$next_functions
+
+```
+
+    [[1]]
+    MulBackward0
+
+```{r}
+# how to compute the gradient for the two branches of y = x * (x + 2),
+# where the left branch is a leaf node (AccumulateGrad for x1)
+out$grad_fn$next_functions[[1]]$next_functions[[1]]$next_functions[[1]]$next_functions
+
+```
+
+    [[1]]
+    torch::autograd::AccumulateGrad
+    [[2]]
+    AddBackward1
+
+```{r}
+# here we arrive at the other leaf node (AccumulateGrad for x2)
+out$grad_fn$next_functions[[1]]$next_functions[[1]]$next_functions[[1]]$next_functions[[2]]$next_functions
+
+```
+
+    [[1]]
+    torch::autograd::AccumulateGrad
+
+If we now call `out$backward()`, all tensors in the graph will have
+their respective gradients calculated.
+
+```{r}
+out$backward()
+
+z$grad
+y$grad
+x2$grad
+x1$grad
+```
+
+    torch_tensor 
+     0.2500  0.2500
+     0.2500  0.2500
+    [ CPUFloatType{2,2} ]
+    torch_tensor 
+     4.6500  4.6500
+     4.6500  4.6500
+    [ CPUFloatType{2,2} ]
+    torch_tensor 
+     18.6000
+    [ CPUFloatType{1} ]
+    torch_tensor 
+     14.4150  14.4150
+     14.4150  14.4150
+    [ CPUFloatType{2,2} ]
+
+After this nerdy excursion, let's see how *autograd* makes our network
+simpler.
+
+## The simple network, now using *autograd*
+
+Thanks to *autograd*, we say good-bye to the tedious, error-prone
+process of coding backpropagation ourselves. A single method call does
+it all: `loss$backward()`.
+
+With `torch` keeping track of operations as required, we don't even have
+to explicitly name the intermediate tensors any more. Just for fun and
+to make the point, we can code forward pass, loss calculation and
+backward pass in just three lines:
+
+```{r}
+y_pred <- x$mm(w1)$add(b1)$clamp(min = 0)$mm(w2)$add(b2)
+  
+loss <- (y_pred - y)$pow(2)$sum()
+
+loss$backward()
+```
+
+Here is the complete code. We're at an intermediate stage: We still
+manually compute the forward pass and the loss, and we still manually
+update the weights. Due to the latter, there is something I need to
```

Comment:
```suggestion
Here is the complete code. We're at an intermediate stage: We still
manually compute the forward pass and the loss, and update the weights. 
Due to the latter, there is something I need to
```

Reply:
"We still manually" only needs to be written the first time, since it applies to both items in the list. 👍

Reply:
Thanks -  I get what you're saying, but I think in this case it is ok to stress this some more (perhaps for psychological reasons: assure people that this will not stay as it is now :-))